File: src/main/java/io/felipeandrade/gsw2/block/ColorProviderOreBlock.java
package io.felipeandrade.gsw2.block;

import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry;
import net.minecraft.block.BlockState;
import net.minecraft.client.color.block.BlockColorProvider;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.BlockRenderView;

import io.felipeandrade.gsw2.material.GSWMaterial;

public class ColorProviderOreBlock extends GSWOreBlock implements BlockColorProvider {

    public ColorProviderOreBlock(GSWMaterial material, Settings settings) {
        super(material, settings);
    }

    @Override
    public int getColor(BlockState state, BlockRenderView view, BlockPos pos, int tintIndex) {
        return material.getPrimaryColor();
    }

    @Override
    public void register() {
        super.register();
        ColorProviderRegistry.BLOCK.register(this, this);
    }
}

File: gsw-mod/src/main/kotlin/io/felipeandrade/gsw/block/ColorProviderOreBlock.kt
package io.felipeandrade.gsw.block

import io.felipeandrade.gsw.material.GSWMaterial
import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry
import net.minecraft.block.BlockState
import net.minecraft.client.color.block.BlockColorProvider
import net.minecraft.client.color.item.ItemColorProvider
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockRenderView

class ColorProviderOreBlock(
    material: GSWMaterial,
    settings: Settings,
) : GSWOreBlock(material, settings), BlockColorProvider, ItemColorProvider {

    override fun getColor(state: BlockState, view: BlockRenderView?, pos: BlockPos?, tintIndex: Int): Int =
        gswMaterial.secondaryColor

    override fun getColor(stack: ItemStack, tintIndex: Int): Int = gswMaterial.secondaryColor

    override fun onClientInit() {
        super.onClientInit()
        ColorProviderRegistry.BLOCK.register(this, this)
        ColorProviderRegistry.ITEM.register(this, this)
    }
}